examples/petsc/multigrid.c

```diff
@@ -47,21 +47,22 @@ int main(int argc, char **argv) {
   char filename[PETSC_MAX_PATH_LEN],
        ceedresource[PETSC_MAX_PATH_LEN] = "/cpu/self";
   double my_rt_start, my_rt, rt_min, rt_max;
-  PetscInt degree = 3, qextra, *lsize, *xlsize, *gsize, dim = 3,
+  PetscInt degree = 3, qextra, *lsize, *xlsize, *gsize, dim = 3, fineLevel,
            melem[3] = {3, 3, 3}, ncompu = 1, numlevels = degree, *leveldegrees;
   PetscScalar *r;
   PetscBool test_mode, benchmark_mode, read_mesh, write_solution;
   DM  *dm, dmOrig;
+  SNES snes_dummy;
   KSP ksp;
   PC pc;
-  Mat *matO, *matI, *matR;
+  Mat *matO, *matPR, matCoarse;
   Vec *X, *Xloc, *mult, rhs, rhsloc;
   UserO *userO;
-  UserIR *userI, *userR;
+  UserProlongRestr *userPR;
   Ceed ceed;
   CeedData *ceeddata;
-  CeedVector rhsceed, diagceed, target;
-  CeedQFunction qf_error, qf_restrict, qf_prolong;
+  CeedVector rhsceed, target;
+  CeedQFunction qf_error, qfRestrict, qfProlong;
```

Comment:
We don't have one strong variable naming convention. I dropped the underscores since QFunctions and Operators are the only libCEED variables with underscores in these examples, but I notice now that they are now the only ones using camel case. We use a mix of snake_case, camelCase, and nocase variable names. Do we want to just pick one and I can convert all of the variables in a separate PR?

Reply:
I guess my comment here was that the QFunction naming convention isn't consistent even on this one line. You've got `qf_error` and `qfRestrict`.

I care more about consistent naming when its scope gets broader. No big deal if it's self-contained to a small function or block, but local consistency is good.

Reply:
Fair point. I skipped over it because I plan on removing it in a follow-up PR that computes the solution on the nodes, but I can add a commit to homogenize in the meantime.

Reply:
Computes the solution on the nodes? Keep in mind that (a) we want to have L^2 error computation and (b) we want to minimize dependence on Lagrange elements (interpolative nodal bases) because we might want to run with other bases (like C^1 continuity).

Reply:
Currently we just have L_infty error computation. In the elasticity code we use the coordinates of the nodes (corners interpolated to the Lobatto points, as usual in our examples) to compute the value of the true solution on the nodes. We use no output basis and scale out the multiplicity of each node when we assemble the global solution vector. Then we use PETSc's built in vector norms on `U - trueSoln` to compute the error.

Reply:
You might mean l_infty. You cannot directly calculate L_infty. You can get at it by a duality estimate from L_1. This is a fragile comparison to use because you can get superconvergence at vertices in many cases.

Reply:
I updated to nocase for this since its what we largely use in the file.

Reply:
It's useful to be able to report
* L^{\infty} at nodes
* L^{\infty} with over-integration (extra quadrature points)
* L^2
* Energy norm of error
* Various norms of flux error

It's fine to not have all of these now, and for some examples to go for simplicity, but I don't think we should delete code that could easily be used to compute these other norms.

Reply:
While you're here, @knepley, is there a simple way we could make `DMCoarsen` (as called by `PCSetUp_MG`) perform p-coarsening?

Reply:
If you mean, can we compute the MG stuff, without having tried it, I think it should work if your coarse mesh is the same Plex with a different DS, at least the computation of the prolongation should work. If you mean, can I make the right thing for p-coarsening, you would need to make another flag (apart from uniform) for this coarsening, and then return the same mesh with a different DS where you decrease the order. May seems to think you should always just slam to order 1 immediately.

Reply:
We want to use our matrix-free prolongation and restriction operators. In the code here, Jeremy has created a hierarchy himself and sets it directly on the PCMG. I'd love to instead set up coarsening callbacks, but would like it to not be an awkward hack.

Going straight to linear elements is usually too aggressive unless you have powerful smoothers.

Reply:
1) It does not look that hacky to me. Its just another coarsening technique where you change the DS instead of the topology. It should only be 20 lines of code I would think.

2) Dave had some good experiments with his spectral code for magma

Reply:
1. Do you envision that as us overriding `dm->ops->createrestriction` or as a new `DMPlex` interface? I'd prefer the latter, but don't know off-hand how it would interact with the rest of your infrastructure.
2. Always fun to compare (can you forward results or should I ask Dave directly?), and to have the capability, but I'm confident that conclusion is not universal.

Reply:
1. So [here](https://gitlab.com/petsc/petsc/-/blob/master/src/dm/impls/plex/plexcoarsen.c#L8) there are two kinds of coarsening: user provided mesh and adaptive coarsening with a label. We could put a third kind in that just changes the DS. At that point we probably want an enum or object, like Refine has now.

2. I would mail Dave directly for that data. I don't think we said it was universal, just works in many more situations than people think.

Reply:
Do you mean that we should do the coarsening ourselves and (make a new interface to) set `dm->coarseMesh`? The link just shows the only call in PETSc to `DMPlexCoarsen_Internal`, which passes `adaptLabel=NULL` (so it being non-NULL is untested).

I wrote Dave.

Reply:
I am not sure what you mean "do coarsening ourselves". I mean there are current two choices: having coarseMesh, and calling DMPlexCoarsen_Internal() (the label is set by an option for this path). We make a flag for a third choice here which just lowers the degree of the FE in the DS.

Reply:
Okay, but `DMPlexCoarsen_Internal` is not callable and there is no code that uses `adaptLabel`. It is passed to `DMPlexLabelToMetricConstraint` which completely ignores that argument. So I don't know how that's supposed to work, but it surely doesn't change any behavior. What public interface do you propose?

Reply:
1) DMPlexCoarsen_Internal() is internal. You are not supposed to be calling it.

2) There definitely is code that uses adaptLabel, but that is not material here.

3) Just like there is a refinement type (Default, Uniform, ToHex, etc) we will add a coarsen type that has p-coarsening, and check it in DMCoarsen_Plex.

Reply:
2. Out of scope, but not in PETSc, and that argument literally isn't used.
```
include/petsc/private/dmpleximpl.h:PETSC_INTERN PetscErrorCode DMPlexCoarsen_Internal(DM, DMLabel, DM *);
src/dm/impls/plex/plexcoarsen.c:  if (!dm->coarseMesh) {ierr = DMPlexCoarsen_Internal(dm, NULL, &dm->coarseMesh);CHKERRQ(ierr);}
src/dm/impls/plex/plexadapt.c:PetscErrorCode DMPlexCoarsen_Internal(DM dm, DMLabel adaptLabel, DM *dmCoarsened)
```

3. So `DMPlexSetCoarsenType(DM, DMPlexCoarsenType)` with `DM_PLEX_COARSEN_H` (current method?), `DM_PLEX_COARSEN_P_UNIFORM`, `DM_PLEX_COARSEN_P_LOG`, `DM_PLEX_COARSEN_P_AGGRESSIVE`? Maybe there should be a parameter in addition to the flag. A PetscObject seems excessive for this.

Reply:
2. If you look inside, I let you override the argument with an option.

3. Yes. I think a parameter rather than flags for all p types.

Reply:
2. The argument is never used in this function. I guess you're saying that although the argument is always ignored (presumably a bug), one can control behavior using the options database. Fine, but that's a different thing.
```c
  ierr = DMPlexLabelToMetricConstraint(dm, adaptLabel, cStart, cEnd, vStart, vEnd, PETSC_DEFAULT, &metricVec);CHKERRQ(ierr);
  ierr = PetscOptionsGetString(NULL, dm->hdr.prefix, "-dm_plex_coarsen_bd_label", bdLabelName, PETSC_MAX_PATH_LEN-1, &flg);CHKERRQ(ierr);
  if (flg) {ierr = DMGetLabel(dm, bdLabelName, &bdLabel);CHKERRQ(ierr);}
  ierr = DMAdaptMetric_Plex(dm, metricVec, bdLabel, dmCoarsened);CHKERRQ(ierr);
```

3. What about two arrays, as in `DMPlexSetCoarsenHierarchy(dm, 5, h_factor={1,1,1,2,2}, ps={4,2,1,1,1})` which would take the original of arbitrary degree (h, p) and reduce to `(h, 4), (h, 2), (h,1), (h/2, 1), (h/4, 1)` where h-coarsening attempts unstructured coarsening with a nominal length scale. Should `h_factor` be real numbers, or even extensible to tensors (for some future anisotropic coarsening)?

Reply:
I think that sounds good. We would just turn those h-factors into inverse eigenvalues for the metric tensor (I think this code is there). I have no good ideas for metric tensor input right now. I think it will take some better tests cases.

Reply:
On second thought, I think I need to leave this incomplete out for now or make it a callback to create the new `PetscFE`. We often want to use overintegration on coarse levels to avoid reevaluating nonlinearities on coarse levels of the linear problem. I doubt I'll have time to do this right for the release.